### ISSUE TITLE
Unsupported modules - torchdp.dp_model_inspector.IncompatibleModuleException

### ISSUE THREAD

**Issue by aleenathomas** [facebookresearch/pytorch-dp#8]
When I run pytorch-dp on a network with embedding layers and linear layers, i get the following exception - torchdp.dp_model_inspector.IncompatibleModuleException. From what I understood from this [post](https://github.com/facebookresearch/pytorch-dp/issues/4), it is thrown when the functionality is not supported. Is there a work-around available to run pytorch-dp for nn.Embedding layers ?

**Comment by Anonymani** [comment]
> When I run pytorch-dp on a network with embedding layers and linear layers, i get the following exception - torchdp.dp_model_inspector.IncompatibleModuleException. From what I understood from this [post](https://github.com/facebookresearch/pytorch-dp/issues/4), it is thrown when the functionality is not supported. Is there a work-around available to run pytorch-dp for nn.Embedding layers ?

Thanks for flagging this.
One option is to get a pre-initialized embedding model (which should work for most use-cases) and then freeze the layer by setting the require_grad attribute of all the layer.parameters() values to False. This will disable back propagation through this layer and things should just work.

If you do need to learn the embedding
At this point unfortunately we do not support back propagation through nn.Embedding as we did not anticipate so many use-cases for it (as pre-trained embedding layers are used for most applications). We do, however, appreciate contributions to our code-base.

**Comment by sergevanhaag** [comment]
I tried implementing a PyTorch ResNet as following:

```
self.cnn = torchvision.models.resnet18(pretrained=True)
self.cnn.fc =  torch.nn.Linear(512, classes)
```

I tried to attach the privacy engine and received the error: 
```
torchdp.dp_model_inspector.IncompatibleModuleException: Model contains incompatible modules.

Some modules are not valid.: ['Main.cnn.bn1', 'Main.cnn.layer3.0.downsample.1']
```

In reality, it gave errors with more than two `bn1` and `downsample` layers, but I cut those out for readability. Is this a similar case?

**Comment by XinyiYS** [comment]
To introduce another case :
`Some modules are not valid.: ['Main.decoder.conv2', 'Main.decoder.conv1']`
Is there a workaround to _conv_ layers or will that be supported later?

**Comment by Liangqiong** [comment]
I also run into the same error for ConvTranspose2d . Interesting to see DCGAN also contains ConvTranspose2d but works.

**Comment by karthikprasad** [comment]
generalizing the issue title as embedding layers are now supported

**Comment by Liangqiong** [comment]
> generalizing the issue title as embedding layers are now supported

Support embedding layers, how about ConvTranspose2d? Is it possible to support this too?

**Comment by karthikprasad** [comment]
Hi @Liangqiong, `ConvTranspose2d` is not supported as of today. The full list of supported layers are at https://github.com/facebookresearch/pytorch-dp/blob/master/torchdp/supported_layers_grad_samplers.py#L124

While supporting GANs is on our list, it's not on the top yet. We do appreciate contributions to our repository.

**Comment by Darktex** [comment]
Let's close this for now as the original question was for nn.Embedding, which did get supported.

@Liangqiong we can certainly support ConvTranspose2d, we just haven't gotten to do it yet. We put it on our list (supporting GANs would be nice indeed!) but if you'd like to help us and take a shot at writing the necessary hooks yourself we would welcome a PR!

**Comment by Antoine-ls** [comment]
> I tried implementing a PyTorch ResNet as following:
> 
> ```
> self.cnn = torchvision.models.resnet18(pretrained=True)
> self.cnn.fc =  torch.nn.Linear(512, classes)
> ```
> 
> I tried to attach the privacy engine and received the error:
> 
> ```
> torchdp.dp_model_inspector.IncompatibleModuleException: Model contains incompatible modules.
> 
> Some modules are not valid.: ['Main.cnn.bn1', 'Main.cnn.layer3.0.downsample.1']
> ```
> 
> These modules were all batch normalization modules. I found that replacing batch normalization with group normalization makes these modules compatible with PyTorch-DP. So if anyone in the future runs into this problem, then use this function which replaces certain layers with others:
> 
> ```
> def convert_layers(model, layer_type_old, layer_type_new, convert_weights=False, num_groups=None):
>     for name, module in reversed(model._modules.items()):
>         if len(list(module.children())) > 0:
>             # Recursion.
>             model._modules[name] = convert_layers(module, layer_type_old, layer_type_new, convert_weights)
>         if type(module) == layer_type_old:
>             layer_old = module
>             # If num_groups is None, GroupNorm turns into InstanceNorm
>             # If num_groups is 1, GroupNorm turns into LayerNorm. 
>             layer_new = layer_type_new(module.num_features if num_groups is None else num_groups, module.num_features, module.eps, module.affine) 
>             if convert_weights:
>                 layer_new.weight = layer_old.weight
>                 layer_new.bias = layer_old.bias
>             model._modules[name] = layer_new
>     return model
> ```
> 
> which you can call with the model that is not working to replace BatchNorm with GroupNorm:
> 
> `model = convert_layers(model, nn.BatchNorm2d, nn.GroupNorm, True, num_groups=2)`
> 
> Hope it helps :)

Hi, your solution is quite helpful, but after replacing bn with gn layers, I encountered this warning message. 

[W accumulate_grad.h:185] Warning: grad and param do not obey the gradient layout contract. This is not an error, but may impair performance.
grad.sizes() = [512, 256, 1, 1], strides() = [256, 1, 131072, 131072]
param.sizes() = [512, 256, 1, 1], strides() = [256, 1, 1, 1] (function operator())

I do not know if this warning is important, but when I try to train the resnet18 with cifar10, the accuracy is always 10%, which means a random classifier, I tried to set sigma to zero, but it did not help, the acc is still 10%, the network is not learning! Could you please tell me if you succeed training with resnet18 in your task? Or if you have any insight, please tell me, thanks in advance~
